### PR TITLE
Make an OSGi bundle

### PR DESCRIPTION
**Pull request by tjwatson**
I would like to use this library as an OSGi bundle.  This PR adds the necessary meta-data to build an OSGi bundle:

- Package `org.checkpoint` is exported as API
- OSGi Bundle-NativeCode header is used to inform the framework about native library
- Try `System.loadLibrary("CheckpointRestore")` first to allow the OSGi class loader to locate the native library packaged in the JAR before doing the manual extraction of the library to the /tmp folder.

This PR also includes adding common patterns to exclude to the `.gitignore`.

It also fixes the RandomTest package name.  This test is in the package directory for org/checkpoint but it was using the default (empty) package name in the source.  While javac and the maven usage of the java compiler cares nothing about the directory `.java` source files are located, I know some IDEs will give errors by default if it detects this type of thing.  I think it would be nice to follow the convention if possible.

### CODE REVIEW

File: JavaCriuJar/pom.xml [hunk 1]
```diff
@@ -119,22 +124,26 @@
...
-                        <phase>package</phase>
-                        <goals>
-                            <goal>shade</goal>
-                        </goals>
-                        <configuration>
-                            <finalName>checkpoint</finalName>
-                        </configuration>
-                    </execution>
-                </executions>
+                <artifactId>maven-jar-plugin</artifactId>
```
Comment (judovana) [inline JavaCriuJar/pom.xml:141]: This change - from sahde to jar plugin, should be in separate PR

File: JavaCriuJar/src/test/java/org/checkpoint/RandomTest.java [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import org.checkpoint.*;
+package org.checkpoint;
```
Comment (judovana) [inline JavaCriuJar/src/test/java/org/checkpoint/RandomTest.java:1]: This shoudl be separate PR.

**Comment by judovana** [comment]
Hi! Pls wait with merging this untill javac -h is merged. As it is now moreover show stopper.
If you can extract the test and sahde->jar hunks to separate PRs, woudl be awesome, as they can be merged asap. (Unlike the rest)

File: JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java [hunk 3]
```diff
@@ -35,7 +38,7 @@
 import java.nio.file.Files;
 import java.nio.file.Paths;
 import java.util.Collections;
-
+@Header(name = "Bundle-NativeCode", value = "libCheckpointRestore.so; osname = Linux; processor = x86-64")
```
Comment (judovana) [inline JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java:41]: Where I have nothing agaisnt osgi integration, I 'm not sure the header is correct. This jar and its native library can be build also on windows, and also on non-intel processors. The original library loading mechanism worked. fine. (tried) The mechanism you propose shoudl work to, but I had not yet tested. But the header.. really looks weird.

File: JavaCriuJar/pom.xml [hunk 4]
```diff
@@ -119,22 +124,26 @@
                     </execution>
                 </executions>
             </plugin>
+			<plugin>
+                <groupId>biz.aQute.bnd</groupId>
+                <artifactId>bnd-maven-plugin</artifactId>
```
Comment (judovana) [inline JavaCriuJar/pom.xml:129]: I'm failing to find any reason to include bnd-maven-plugin. it brings no help. If it do, by my unseen way, please, fix formating!

File: .gitignore [hunk 5]
```diff
@@ -1,2 +1,42 @@
 /JavaCriuJar/target/
 /JavaCriuJar/nbproject/
+*#
```
Comment (judovana) [inline .gitignore:3]: Actually also this deserves its own PR,  as it amy be merged immediately.

File: JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java (hunk 3 above)
Comment (tjwatson) [reply JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java:41]: I was not aware of support for CRIU on anything but Linux.  Do you have documentation that describes using CRIU on Windows?  

If other processors are to be supported by a single JAR then I assume multiple native libraries will need to be packaged in the same JAR.  In that case both the OSGi header for Bundle-NativeCode and the manual copying and loading of the extracted library will need to be updated to pick the proper library for the processor.

If instead the build is going to produce multiple JARs (one per supported processor) then we likely could drop the processor attribute altogether.

File: JavaCriuJar/pom.xml (hunk 4 above)
Comment (tjwatson) [reply JavaCriuJar/pom.xml:129]: This is what processes the annotations:

`org.osgi.annotation.bundle.Header`
`org.osgi.annotation.bundle.Export`
`org.osgi.annotation.versioning.Version`

It is also what creates the valid bundle manifest META-INF/MANIFEST.MF.  Without using this plugin generating a bundle manifest is an arduous task.

I'll fix the formating.

File: JavaCriuJar/pom.xml (hunk 1 above)
Comment (tjwatson) [reply JavaCriuJar/pom.xml:141]: This goes hand in hand with the use of bnd-maven-plugin in order to make sure the generated manifest gets pulled into the JAR.  I could not see any reason the shade plugin was being used before.

File: JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java (hunk 3 above)
Comment (tjwatson) [reply JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java]: I've removed the processor attribute for now.  Once there is a multi-platform build strategy we can revisit the native code selection.

File: JavaCriuJar/pom.xml (hunk 4 above)
Comment (tjwatson) [reply JavaCriuJar/pom.xml:126]: I've fixed the formatting.

File: .gitignore (hunk 5 above)
Comment (tjwatson) [reply .gitignore]: I've done this with #12

File: JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java (hunk 3 above)
Comment (judovana) [reply JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java]: hi! You are of course right, and windows is out of game.  My wrong above.  The multi architecture however remains in game. Thank you for removing the  x86-64. It is ok by  me now.

File: JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java [hunk 6]
```diff
@@ -159,18 +162,22 @@ private static void copyLibrary(String library, String destDir) throws IOExcepti
...
-        if (!Files.exists(Paths.get(tmpLib))) {
-            try {
-                copyLibrary("CheckpointRestore", libDir);
-            } catch (IOException ex) {
-                throw new RuntimeException("can't initialize CheckpointRestore",ex);
+
+        try {
+        	System.loadLibrary("CheckpointRestore");
+        } catch (UnsatisfiedLinkError e) {
+            String libDir = "/tmp";
```
Comment (judovana) [inline JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java:169]: I kow ths issue was present also in original code, but shouldn't   System.getProperty("java.io.tmpdir"); be used?

File: JavaCriuJar/pom.xml (hunk 4 above)
Comment (judovana) [reply JavaCriuJar/pom.xml:126]: ok.

File: JavaCriuJar/pom.xml (hunk 1 above)
Comment (judovana) [reply JavaCriuJar/pom.xml:138]: Sure. The usage of jar over shade is definitely good thing.
Thanx for highligting the conenction to the bnd-maven-plugin

File: JavaCriuJar/pom.xml (hunk 4 above)
Comment (judovana) [reply JavaCriuJar/pom.xml:126]: ty!

File: .gitignore (hunk 5 above)
Comment (judovana) [reply .gitignore]: ty!

File: JavaCriuJar/pom.xml [hunk 7]
```diff
@@ -45,12 +45,17 @@
...
+            <scope>compile</scope>
+        </dependency>
     </dependencies>
 
-    <build>
-        <plugins>
-            <plugin>
+	<build>
+		<plugins>
+			<plugin>
```
Comment (judovana) [inline JavaCriuJar/pom.xml:58]: There is still some weird formatting.

File: JavaCriuJar/pom.xml (hunk 1 above)
Comment (judovana) [reply JavaCriuJar/pom.xml:138]: And sorry for nitpicking about bnd-maven-plugin . My imagination was distribution-wild-wide and first thought I had was "ouch.. that will be fun to patch it out again".  Except the very minor formatting issue still present in pom, I'm happy with this. But Chf must say a final word whether to include  osgi or not.

File: JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java (hunk 6 above)
Comment (tjwatson) [reply JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java:169]: Probably, but I also think this extraction of the library to temp thing will need to be reworked altogether to be reliable.  I plan to use this in OSGi where the native code extraction and management on disk is handled by the framework so I will not run into these issues myself.  For example, it does nothing to see if the existing library is different than the one packaged in the jar.  If an old library is at the tmp location it just tries to use it.

Comment (judovana) [reply JavaCriuJar/src/main/java/org/checkpoint/CheckpointRestore.java:169]: fair enough. TY!

File: JavaCriuJar/pom.xml (hunk 7 above)
Comment (tjwatson) [reply JavaCriuJar/pom.xml]: Fixed now
